cordapp-example/workflows-kotlin/src/main/kotlin/com/example/flow/PaybackFlow.kt

```diff
@@ -53,76 +66,88 @@ object PaybackFlow {
                     .addInputState(inputStateAndRef)
                     .addCommand(txCommand)
 
-            // Stage 2.
             progressTracker.currentStep = VERIFYING_TRANSACTION
             // Verify that the transaction is valid.
             txBuilder.verify(serviceHub)
 
-            // Stage 3.
             progressTracker.currentStep = SIGNING_TRANSACTION
-            // Sign the transaction.
-            val signedTx = serviceHub.signInitialTransaction(txBuilder, inputIouState.lender.owningKey)
-
-            /**
-             * I don't think lender sig is required here,
-             * we should be able to payback without it - but idk how to do `FinalityFlow` without it
-             * also, will it update lender vault and spend IUO if lender will nobe be included in flow?
-             */
+            // Sign the transaction (must be by a borrower)
+            val signedTx = serviceHub.signInitialTransaction(txBuilder)
 
+            progressTracker.currentStep = INITIALIZE_OTHER_PARTY_FLOW
             // Send the state to the counterparty, and receive it back with their signature.
-            val otherPartySession = initiateFlow(inputIouState.lender)
-            //val fullySignedTx = subFlow(CollectSignaturesFlow(signedTx, setOf(otherPartySession), ExampleFlow.Initiator.Companion.GATHERING_SIGS.childProgressTracker()))
+            val lenderPartySession = initiateFlow(inputIOUState.lender)
+
+            val confirmPayback = lenderPartySession
+                    .sendAndReceive<Boolean>("ready for receiving payback?")
+                    .unwrap { it }
+
+            require(confirmPayback) { "Oops, lender failed to be ready" }
+
+            lenderPartySession.send(paybackValue) // */
+
+            progressTracker.currentStep = GATHERING_SIGS
+            val fullySignedTx = subFlow(CollectSignaturesFlow(
+                    signedTx,
+                    setOf(lenderPartySession),
+                    ExampleFlow.Initiator.Companion.GATHERING_SIGS.childProgressTracker())
+            )
 
-            // Stage 4.
             progressTracker.currentStep = FINALISING_TRANSACTION
             // Notarise and record the transaction in both parties' vaults.
-            return subFlow(FinalityFlow(signedTx, setOf(otherPartySession), FINALISING_TRANSACTION.childProgressTracker()))
+            return subFlow(FinalityFlow(fullySignedTx, setOf(lenderPartySession), FINALISING_TRANSACTION.childProgressTracker()))
         }
     }
 
     @InitiatedBy(Initiator::class)
-    class Acceptor(val otherPartySession: FlowSession) : FlowLogic<SignedTransaction>() {
+    class Acceptor(val borrowerPartySession: FlowSession) : FlowLogic<SignedTransaction>() {
         @Suspendable
         override fun call(): SignedTransaction {
-            val signTransactionFlow = object : SignTransactionFlow(otherPartySession) {
+            //send confirmation about being ready to receive payback
+            borrowerPartySession.send(true)
+
```

Comment:
The initiator calls `lenderSession.send(paybackValue)` before calling `CollectSignaturesFlow` so your `receive` and `signTransaction` should appear in the same order, meaning:   
1. First `receive` the payback value and validate that it matches the IOU value.
2. Then sign the transaction.   

Again order in send and receive is very important!!!